e2e-tests/tests/enketo.spec.js

```diff
@@ -113,6 +106,85 @@ test.describe('Enketo', () => {
     });
   });
 
+  test.describe('offline form', () => {
+    const offlineUrls = [{
+      description: 'Offline Submissions',
+      url: ({ enketoId }) => `/-/x/${enketoId}`, requireLogin: true,
+      newUrl: ({ enketoId }) => `/-/x/${enketoId}`
+    }, {
+      description: 'Offline Submissions Public Link',
+      url: ({ enketoId, st }) => `/-/x/${enketoId}?st=${st}`, requireLogin: false,
+      newUrl: ({ enketoId, st }) => `/-/x/${enketoId}?st=${st}`
+    }];
+
+    offlineUrls.forEach(t => {
+      test(`renders offline Form - ${t.url}`, async ({ page }) => {
+        const { enketoId, enketoOnceId, xmlFormId } = publishedForm;
+        const { enketoId: draftEnketoId } = draftForm;
+        const { instanceId } = firstSubmission;
+        const { token: st } = publicLink;
+
+        if (t.requireLogin) {
+          await login(page);
+        }
+
+        await page.goto(appUrl + t.url({ enketoId, enketoOnceId, draftEnketoId, xmlFormId, instanceId, st }));
+
+        await page.waitForURL(appUrl + t.newUrl({ enketoId, enketoOnceId, draftEnketoId, xmlFormId, instanceId, st }));
+
+        await expect(page.getByRole('heading', { name: publishedForm.name })).toBeVisible();
+
+        await expect(page.getByText('A new version of this application has been downloaded. Refresh this page to load the updated version.')).toBeVisible();
+      });
+    });
+
+    const rejectAllRequests = route => { route.abort('failed'); };
+
+    const goOffline = async (page) => {
+      await page.context().setOffline(true);
+      await page.route('**', rejectAllRequests);
```

Comment:
Rejecting all request via page.route is necessary here. I was assuming that setting offline to true would block everything but that's not the case. 

Requests via service workers still go through + `connection: keep-alive` and caching also cause requests to succeed.